config-controller/pkg/client/client.go

```diff
@@ -0,0 +1,303 @@
+
+	svc := v1.NewPolicyServiceClient(conn)
+	perRPCCreds.svc = v1.NewAuthServiceClient(conn)
+
+	return &grpcClient{
+		perRPCCreds: perRPCCreds,
+		svc:         svc,
+	}, nil
+}
+
+func (gc *grpcClient) ListPolicies(ctx context.Context) ([]*storage.ListPolicy, error) {
+	allPolicies, err := gc.svc.ListPolicies(ctx, &v1.RawQuery{})
+	if err != nil {
+		return []*storage.ListPolicy{}, errors.Wrap(err, "Failed to list policies from grpc client")
+	}
+
+	return allPolicies.Policies, err
+}
+
+func (gc *grpcClient) GetPolicy(ctx context.Context, id string) (*storage.Policy, error) {
+	policy, err := gc.svc.GetPolicy(ctx, &v1.ResourceByID{Id: id})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to fetch policy %s", id)
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) PostPolicy(ctx context.Context, policy *storage.Policy) (*storage.Policy, error) {
+	req := &v1.PostPolicyRequest{
+		Policy:                 policy,
+		EnableStrictValidation: true,
+	}
+
+	policy, err := gc.svc.PostPolicy(ctx, req)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to POST policy")
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) PutPolicy(ctx context.Context, policy *storage.Policy) error {
+	_, err := gc.svc.PutPolicy(ctx, policy)
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to PUT policy")
+	}
+
+	return nil
+}
+
+func (gc *grpcClient) TokenExchange(ctx context.Context) error {
+	if time.Since(gc.perRPCCreds.lastUpdated).Seconds() > 60.0 {
+		return gc.perRPCCreds.refreshToken(ctx)
+	}
+	return nil
+}
+
+type client struct {
+	svc         PolicyClient
+	cache       map[string]*storage.Policy
+	lastUpdated time.Time
+}
+
+type clientOptions interface {
+	Apply(CachedPolicyClient)
+}
+
+func New(ctx context.Context, opts ...clientOptions) (CachedPolicyClient, error) {
+	c := client{}
+
+	for _, o := range opts {
+		o.Apply(&c)
+	}
+
+	if c.svc == nil {
+		var gc PolicyClient
+		err := retry.WithRetry(func() error {
+			var innerErr error
+			gc, innerErr = newGrpcClient(ctx)
+			if innerErr != nil {
+				getLogger(ctx).Error(innerErr, "Failed to connect to Central")
+			}
+
+			c.svc = gc
+
+			if innerErr = c.EnsureFresh(ctx); innerErr != nil {
+				getLogger(ctx).Error(innerErr, "Failed to initialize client")
+			}
+
+			return innerErr
+		}, retry.Tries(10), retry.WithExponentialBackoff())
+
+		if err != nil {
+			return nil, errors.Wrap(err, "could not initialize policy client")
+		}
+	} else {
+		if err := c.EnsureFresh(ctx); err != nil {
+			getLogger(ctx).Error(err, "Failed to initialize client")
+		}
+	}
+
+	return &c, nil
+}
+
+func (c *client) ListPolicies(ctx context.Context) ([]*storage.Policy, error) {
+	policies := make([]*storage.Policy, 0, len(c.cache))
+	for _, value := range c.cache {
+		policies = append(policies, value)
+	}
+	return policies, nil
+}
+
+func (c *client) GetPolicy(ctx context.Context, name string) (*storage.Policy, bool, error) {
+	policy, exists := c.cache[name]
+	return policy, exists, nil
```

Comment:
nit: return c.cache[name], nil (I _think_ that should work)

Reply:
sadly not 
![image](https://github.com/user-attachments/assets/63856c28-79a3-4ed1-b356-1ea94570f8b9)